Increased Configurability and Edge Case Handling

KritR opened the pull request:
Lots of changes here. The goal of this PR is to add configurable parameters to numerizer and to increase the handling of edge cases through context. #12 

First are the two options for configuration. 
The bias parameter allows for two biases, towards :fractional or :ordinal. The default is :none.
In ambiguous situations, it assumes bias towards fractions or ordinals depending on the string. There may be some fine tweaking required with numerizer's behavior in these configurations. Additional tests may make these clear.```
The second parameter is the ignore parameter, allowing users to specify a set of strings to ignore the parsing of. In the case of Chronic, this could potentially be useful for items like 'seconds'. However, at present, all of Chronic's tests pass for this branch of numerizer, so it will have to be seen.

Next are the additional edge cases detected.
In the ambiguous mode it is as follows
- ````second```` is now parsed as an ordinal
 - ````the```` preceding ord/fracs will be assumed to be ordinals
-  in the event of seconds, an ordinal preceeding ````second```` will prevent the ````second```` from also being parsed
- a pronoun preceding ````quarter```` will cause quarter to not be parsed ( assuming verb tense )
- fractions following ordinals should now properly work and hanging fractions are cleaned up
- ````half```` is now substituted in for its fractional variant by default 
- ````halves```` is now also parsed

In order to make some of these things possible, and to clean things up. The ordinals and fractionals have now merged where common into a ````ambiguous_ordinal_fractionals```` constant. In addition, there are some other precomputed constants like the regex for the ordinals and the pluralized fractions.

Lastly, couple tests have been rearranged and new ones have been added to check if added features are working. All tests appear to pass except the one commented out for future solving.

jduff commented on test/test_numerizer.rb:
```diff
@@ -127,4 +131,42 @@ def test_ordinal_strings
+  end
+
+  def test_bias_ordinal
+    assert_equal '4th', Numerizer.numerize('fourth', bias: :ordinal)
```
Will `a fourth of` numerize to `a 4th of` when bias is ordinal?

KritR replied:
It should be added. Currently it results in "1 4th of" instead of "a 4th of".

KritR replied:
it'll result in "a 4th of" in the latest commit.

jduff commented on test/test_numerizer.rb:
```diff
@@ -95,7 +99,7 @@ def test_compatability
     {
       'first' => '1st',
-      'second' => 'second',
+      'second' => '2nd',
```
👍

jduff commented on lib/numerizer/version.rb:
```diff
@@ -1,3 +1,3 @@
 class Numerizer
-  VERSION = '0.2.0'
+  VERSION = '0.3.0'
```
Version bump should happen before release, not with the changes

jduff commented on lib/numerizer.rb:
```diff
@@ -108,50 +105,105 @@ class Numerizer
     andition(string)
 
+    # always substitute halfs
+    unless ignore.include? 'half' then
```
You don't need the `then`, you have this in a few places and it can be dropped in all of them.

jduff commented on lib/numerizer.rb:
```diff
@@ -108,50 +105,105 @@ class Numerizer
+    ALL_ORDINALS.each do |on|
+      break if bias == :fractional # shouldn't be necessary but saves cycles
+      next if ignore.include? on[0].downcase 
+      if on[0] == 'second' and bias != :ordinal then
```
I prefer `&&` instead of `and`. `and` has a lower precedence which can cause some odd behaviour and should usually be avoided except in very special circumstances. More [here](https://stackoverflow.com/questions/1426826/difference-between-and-and-in-ruby) if you're interested.

jduff commented on lib/numerizer.rb:
```diff
@@ -108,50 +105,105 @@ class Numerizer
-    (DIRECT_ORDINALS + SINGLE_ORDINALS).each do |on|
-      string.gsub!(/(^|\W)#{on[0]}(?=$|\W)/i, '\1<num>' + on[1].to_s + on[0][-2, 2])
+    ALL_ORDINALS.each do |on|
+      break if bias == :fractional # shouldn't be necessary but saves cycles
```
Why not skip entering into the loop instead of breaking once you are in there?

jduff commented on lib/numerizer.rb:
```diff
@@ -108,50 +105,105 @@ class Numerizer
+        string.gsub!(/(^|\W)#{tp[0]}(?=$|\W)/i, '/' + tp[1].to_s)
+      elsif /quarter/ =~ tp[0] then
+        string.gsub!(/(?:\w*)(^|\W)#{tp[0]}(?=$|\W)/i) do |match|
+          if (match =~ /^(i|you|he|she|we|it|you|they|to|the)/i) == nil then 
```
This regexp should probably be a constant. A comment describing what is happening here might be good as well (if the constant is named well that could cover it).

jduff commented on lib/numerizer.rb:
```diff
@@ -108,50 +105,105 @@ class Numerizer
+    end
+
+    fractionals.each do |tp|
+      next unless ignore.select {|x| /#{tp[0]}/ =~ x } .empty?
```
Using `none?` or `any?` instead of `select` with `empty?` might be a bit simpler

jduff commented on lib/numerizer.rb:
```diff
@@ -108,50 +105,105 @@ class Numerizer
     TEN_PREFIXES.each do |tp|
+      next if ignore.include? tp[0].downcase 
       SINGLE_NUMS.each do |dn|
+        next if ignore.include? dn[0].downcase 
```
I'm not a fan of these ignore checks all over the place, it seems easy to miss one if the code changes. I don't have a great suggestion though :/.

It also adds a lot of extra loops to run through which isn't super efficient.

jduff commented:
I like where you're going with this, but it does make the code a bit more complicated. I think a few more tests would be good (I don't think I noticed any that matched the issue from chronic duration) and some more eyes to review (maybe @davispuh wouldn't mind since he's helped out here before?)

Thanks for taking a stab at this, not an easy task by any means!

davispuh commented:
With quick look it seems fine, but it does make code more harder to read and there's already so much Regexp. I think it would be good idea to split it in more methods.

Also I suspect that using proper parsing would have better performance than `gsub` and code would be easier to read. With this I mean iterate source string and build target, for example using `StringScanner` so we would process each source string part just once, rather than now each `gsub` goes over whole string every time.

Another thing to think about is multiple language support. In future I might be interested in numerizing other languages too not just English so it would be nice if there were already some abstraction.
Maybe something like this
```ruby
class GenericNumerizer
  def numerize(str)
    numerize_ordinals(str)
  end

  def numerize_ordinals(str)
    raise 'must be implemented in subclass'
  end
end

class EnglishProvider < GenericNumerizer
  def numerize_ordinals(str)
  # ...
  end
end

class Numerizer
  PROVIDERS = { 'en' =>  EnglishProvider.new }
  def self.numerize(string, langauge = 'en')
    PROVIDERS[langauge].numerize(string)
  end
end
```

KritR commented:
I appreciate the in depth review comments! I'll try and refactor this into something more efficient and comprehensible. May be a week or two before I get around to this.
I figured that the regex approach would work at first, but with more complicated checks, it only ended up nesting gsubs.
I'll try and incorporate the multilingual support you outlined @davispuh.

KritR commented:
Sorry about the delay.

Update on where things are at, I removed the looping over gsubs through the usage of hashes. Just running the test cases, it's gone from around 0.65 seconds to 0.15 seconds. So around a 4x improvement with that many cases. So it appears that's one way to mitigate some performance issues.  I also removed the unnecessary nested gsubs, which also reduces the complexity of the code a little bit.

I wasn't sure how to implement the single scan parsing. The order of the parsing in this matters and it wasn't obvious how that would work. Maybe storing the scan data and then operating over it backwards so that string positions aren't messed up during the operations ? I think we'll still need 4 scans for each sequential operation occurring.

I also modified the Numerizer class to now use the language providers just as @davispuh suggested. 

Still have to add more tests, if you have any tests in mind please mention them or just make a pull request to my repo.

davispuh commented:
Looks really good, great work! 👍

KritR commented:
I can't think of any other tests to add at this point. Thoughts @jduff

jduff commented on test/test_numerizer_en.rb:
```diff
@@ -127,4 +139,48 @@ def test_ordinal_strings
+    assert_equal '1/4', Numerizer.numerize('fourth', bias: :fractional)
+    assert_equal '1/12', Numerizer.numerize('twelfth', bias: :fractional)
+    assert_equal 'the 1/4', Numerizer.numerize('the fourth', bias: :fractional)
+    assert_equal '2.75', Numerizer.numerize('two and three fourths', bias: :fractional)
```
It would probably be good to have the same cases between ordinal and fractional bias so you can easily see the differences

KritR commented on lib/providers/english_provider.rb:
```diff
@@ -0,0 +1,214 @@
+        fraction = $3
+        # HANDLES QUARTER EDGE CASES
+        if (!$3.start_with?('quarter') && space == ' ') || ($3.start_with?('quarter') && (prematch =~ /^(#{PRONOUNS})\b/) == nil) 
+            match = prematch << '/' << ALL_FRACTIONS[fraction].to_s 
```
Fewer spaces within the conditionals or remove the extra space after the conditionals ?

jduff replied:
Sorry, it just looks like there is two extra spaces before `match`

KritR commented:
Those two commits should fix all the previous concerns.

KritR commented:
The API should be backwards compatible. However, it might make a difference for use cases like Chronic where it was assumed 'second' would not be parsed.

davispuh commented:
It's a lot of changes so even if it is backwards compatible I think it does deserve `1.0.0` release.

KritR commented:
Any updates @jduff ?

jduff commented:
Sorry for not getting back to this sooner!

Everything looks good here, I just want to double check running the tests when I’m at my computer again before merging. Should have it merged sometime next week.

Thanks again for all your work on this, really great stuff!

jduff commented on test/test_numerizer_en.rb:
```diff
@@ -24,7 +31,7 @@ def test_straight_parsing
       100 => ['one hundred', 'a hundred', 'hundred a'],
       150 => 'one hundred and fifty',
-      # 150 => 'one fifty',
+      150 => 'one fifty',
```
just noticed that this uses the same key as the one above so only one is kept in the hash, you need to put the two two cases into an array like we have for the `100` case.
